externalcatalog: add DropIngestedExternalCatalog

msbutler opened the pull request:
This patch introduces a new api that LDR's OnFailOrCancel routine will use to
drop created table descriptors if the user's LDR job fails or gets cancelled
during an offline initial scan.

Epic: none

Release note: none

msbutler committed:
externalcatalog: add DropIngestedExternalCatalog

This patch introduces a new api that LDR's OnFailOrCancel routine will use to
drop created table descriptors if the user's LDR job fails or gets cancelled
during an offline initial scan.

Epic: none

Release note: none

msbutler committed:
externalcatalog: move SetGCTTLForDroppingTable to externalcatalog

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138107)
<!-- Reviewable:end -->

msbutler committed:
crosscluster/logical: drop tables after failed/cancelled offline scan

This patch also plumbs the ingested external catalog to the job record, to
enable rolling back and publishing tables, and in the future, child
descriptors.

Epic: none

Release note: none

fqazi reviewed (changes requested):
> Reviewed 4 of 4 files at r1, 4 of 4 files at r2, 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138107)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @msbutler)

___
*[`pkg/sql/catalog/externalcatalog/external_catalog.go` line 191 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/138107#-OFwN2im9TVS9leulYkr:-OFwN2im9TVS9leulYks:b-xecf6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d543ab6aba20b67b3cdfdf5a4e794008bbd73e9b/pkg/sql/catalog/externalcatalog/external_catalog.go#L191)):*
> ```Go
> 		// Drop any schedules we may have implicitly created.
> 		if tableToDrop.HasRowLevelTTL() {
> 			scheduleID := tableToDrop.RowLevelTTL.ScheduleID
> ```

This seems a bit precarious? Should we extend the descriptor interface to get all scheduled jobs? I know its only one today


<!-- Sent from Reviewable.io -->

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138107)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @fqazi)

___
*[`pkg/sql/catalog/externalcatalog/external_catalog.go` line 191 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/138107#-OFwN2im9TVS9leulYkr:-OFwkQ-75_mTIXkGaxKe:be4vvxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d543ab6aba20b67b3cdfdf5a4e794008bbd73e9b/pkg/sql/catalog/externalcatalog/external_catalog.go#L191)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

This seems a bit precarious? Should we extend the descriptor interface to get all scheduled jobs? I know its only one today
</blockquote></details>

This bit is a pure copy and past for what is currently done in restore. Could we consider extending the descriptor interface in a follow up pr?


<!-- Sent from Reviewable.io -->

dt reviewed (approved):
last commit LGTM (leaving first two to foundations review)

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138107#-:-OFx-caN5861RyifKZ5H:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @msbutler)

___
*[`pkg/sql/catalog/externalcatalog/external_catalog.go` line 191 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/138107#-OFwN2im9TVS9leulYkr:-OFx-a4n7N9C-mawKwTh:b-9fe0v6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d543ab6aba20b67b3cdfdf5a4e794008bbd73e9b/pkg/sql/catalog/externalcatalog/external_catalog.go#L191)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

This bit is a pure copy and past for what is currently done in restore. Could we consider extending the descriptor interface in a follow up pr?
</blockquote></details>

Yes, lets leave this for a future PR.


<!-- Sent from Reviewable.io -->

msbutler commented:
TFTR!

bors r=fqazi
